### ISSUE TITLE
Numerous issues with lamp parameters

### ISSUE THREAD

**Issue by jdent02** [bsavery/PRMan-for-Blender#152]
So I was attempting to try out some of the other lamps in PRMan aside from the rectangle/square area light and I ran into some serious issues in which the settings didn't work as expected or at all.  This is a lengthy one with numerous pictures......

With the light tab selection set to 'area' I get the following result with every 'shape' setting in the PxrStdAreaLight node aside from 'spot':
![area-cylinder](https://cloud.githubusercontent.com/assets/13521172/11017996/c846c3ec-8579-11e5-909f-4826b3f044ab.png)

With the 'spot' shape selected I get this:
![area-spot](https://cloud.githubusercontent.com/assets/13521172/11017999/d54068be-8579-11e5-8dab-c5da796e8696.png)

With the light tab selection set to 'spot' and the node shape set to 'spot' I get typical spot behavior:
![spot-spot](https://cloud.githubusercontent.com/assets/13521172/11018005/f2c9ea04-8579-11e5-85a1-52e538e79a04.png)

However the spot tab setting and every other shape node settings gives me this:
![spot-cylinder](https://cloud.githubusercontent.com/assets/13521172/11018008/07746e0c-857a-11e5-8022-5b3957293999.png)  This appears to be a disk light.

For the 'point' light tab setting everything gives me this:
![point-distant](https://cloud.githubusercontent.com/assets/13521172/11018012/29d5252c-857a-11e5-9b35-b53804d7a3bd.png)  This appears to be a sphere light.

Except the 'spot' shape, which seems to give me a spotlight.

So to summarize the 'distant' and 'cylinder' shapes aren't accessible in any way at this point.  Sphere and disk shapes are available, but only by using non intuitive combinations of light settings from both the Blender light tab and the PxrStdAreaLight node settings.  Also, for the sphere and disk lights the only way to set their size is by scaling the light object.

**Comment by bsavery** [comment]
Thanks for the detailed report.  Generally currently the "shape" on the shader controls the emission profile while the tabbed setting controls the shape.  What we should really do is:

1  Make the tab setting control the light geo type, 
2.  when in area mode, let the shape parameter on the shader drive the geo type. (rect|sphere|cylinder etc)
3.  Drive the disk/cylinder width/length off the size there.  Also with the sphere.

(as a note there's no real "point" light in RIS mode.  The best we can do is a really small sphere)

**Comment by jdent02** [comment]
That's how I thought it was going to work originally.

However, I'm not sure if you'll be able to get that kind of system working with the way the lights use the node system currently.

Currently the PxrStdAreaLight is a node just like any other, and you can select options in there independently of what is selected in the light tab (as far as the viewport shape).  In Cycles the actual light type parameters (area, spot, point) are not exposed in the node tree and the only thing you can use nodes for is setting intensity and color.  Maybe the PRMan lights should be handled in a similar way?  As in remove the PxrStdAreaLight node (and the other ones) from the node tree.  If a user selects 'spot' on the light tab, behind the scenes an area light is added to the scene and only the parameters applicable to a spot are presented to the user *in the light tab* section.  All those parameters (cone size, shadow options) would be set in this section, and the node tree would only allow influence to the color, intensity or image mapping inputs.

**Comment by bsavery** [comment]
Agreed.  I'm actually coding that same thing up as week speak.  Actually we're sort of half in that world in that if you chose area light it automagically hooks up the PxrStdAreaLight Node.  So I was already thinking of doing what you said.  I'm going to hide the shape from the shader node and move that to the light tab section along with the size like you said.  Great minds think alike!

**Comment by jdent02** [comment]
If my mind was great I'd be able to help out with the code and not just armchair quarterback.......:)

**Comment by bsavery** [comment]
Try the latest fix.  

https://www.dropbox.com/s/nja3eeqy4gwbpsc/light_test.blend?dl=0

**Comment by jdent02** [comment]
Gave it a test.  Most things work great now, very nice work!  The only glitches I noticed were:

Angular visibility doesn't seem to work for the distant lamp.  I noticed the parameter wasn't in the light tab but still in the node.  Also changing the light exposure, or color will trigger an IT refresh but the image doesn't change, so you have to restart the interactive session to see any updates (material updates work as expected though).

Aside from that I have a few minor suggestions:

I'd remove the point light option entirely.  As you mentioned it's not even a valid option for RIS.  Also, I'd remove the Cone and Penumbra spot settings from the PxrStdAreaLight node since they appear to be locked to the spot settings on the light tab (as they should be) and even if you change them in the node they automatically align with the tab settings when you hit the render button.  Also I'd move the Angular Visibility for the distant light to the light tab and remove that from the node as well, that would make it nice and consistent with everything else.

That's all I noticed.  For such a big update it seems pretty free of serious bugs.

**Comment by bsavery** [comment]
I would agree about removing point light options.  With one caveat.  If you don't click that "Add renderman nodetree option" we should still translate point lights.  Point being that if someone brings in a scene from blender (that has point lights) we want to translate them to something.  After they click that nodetree button which maybe should be changed to "Make this light a RenderMan physical light" or something, there's no need for point lights really.

**Comment by jdent02** [comment]
Or maybe you could automatically translate them into very small sphere lights and trigger a warning popup about it......?

On a side note I noticed that the Blender interface no longer locks up completely when launching an IR session.  That's nice!  It still does when you try to shut it down during a refresh, but I imagine that's a Blender problem.

**Comment by bsavery** [comment]
I'll have to ask the develop who wrote the shader about angular visibility tomorrow.  Not sure that works at all.

**Comment by jdent02** [comment]
Okay.  BTW, my note about the IT refresh not showing lighting changes needs to be amended.  It works just fine, given that you delete and reintroduce any lights that existed in the scene prior to the latest updates.

**Comment by bsavery** [comment]
While I have your attention... On the subject of environment lights.  It seems like the "blender way" is to have it in the world settings.  What do you think of putting a built in env map light in the world settings to match up with this idea better?  People can always add a second manually if they want.

**Comment by MWY3510** [comment]
I agree that using the world settings as the main environment light would be the "blender way." By the way great job on the recent fixes!
MWY3510

**Comment by jdent02** [comment]
Yeah the world panel is the default place for setting up environment lighting in Cycles and BI.  Would you plan to put both the Skylight and Environment light there?  Also Cycles doesn't give you a straightforward way in the world tab to rotate the environment map, so it would be a nice plus if PRMan did.

**Comment by RonFrancis** [comment]
Just a note about distant light sources, (emulating sunlight).

Note: This based on the current version from here 
https://github.com/bsavery/PRMan-for-Blender/
I was unable to download the dropbox link above and don't know what it is.

Distant light is currently being attenuated from the object position rather than it being at an infinite distance, and shadows are also calculated from the object position rather than being parallel.

Here is an example where the light was a 'point' and the shape was 'distant'.
![distant light](https://cloud.githubusercontent.com/assets/15881747/11203167/3502a82a-8d45-11e5-9ffc-373fd6b72b83.jpg) 
I'm pretty new to this stuff, so I hope I didn't miss something obvious.

By the way, thanks for this plug-in. Fantastic.

**Comment by jdent02** [comment]
You sure you have the latest version?  You shouldn't be able to select any kind of shape with the point light selected. The distant lamp should be an option for the area light

**Comment by RonFrancis** [comment]
Much less sure now!
I certainly downloaded it from the link above yesterday and put it in my plugins folder, but I guess it's possible it didn't overwrite the previous files because I hit the wrong button or something.
I'll try again. Cheers.

**Comment by jdent02** [comment]
I think I had that problem before so I delete the entire PRMan for Blender folder from the add ons before copying the new one over.

Under the current revision you should be able to select Area, Spot, Point, Distant, Sky or Environment light from the main light object info tab.  The shape parameter has been removed from the PxrStdAreaLight node.

**Comment by RonFrancis** [comment]
Yes, that was the problem and it seems to be working well now. 
I thought I knew my way around computers a bit better than that! :(
Thanks, much appreciated.

**Comment by jdent02** [comment]
Got another glitch.  The blend slider for the spot light works in the opposite way from what's shown in Blender's 3d view.  Increasing the blend causes the light penumbra to expand outwards, where in the 3D view it shrinks inwards.

**Comment by bsavery** [comment]
I think that has to do with the way renderman expects the penumbra to go vs the way blender does.... it's possible just throwing a negative in there will make it match up.

**Comment by bsavery** [comment]
Actually I think this might have more to do with the "size" of the spot light.  In blender it appears to be an infinitesimally small spot light, whereas we make it as a disk with actual area.

**Comment by jdent02** [comment]
Yes, under normal settings the spot is emitted from a single point in space.  Cycles allows you to optionally give it a physical size but that setting is different than the cone angle setting that is currently exposed for the exporter.

**Comment by jdent02** [comment]
Hey Brian, I noticed this was still open.  Aside from the distant lamp angular visibility parameter (which I  haven't tested lately so it may be fixed for all I know) I think everything else has been seen to so you can probably close this and free up some more space on the issues page.
